File: Python3/medium/AddTwoNumbers.py
from object.ListNode import ListNode


class AddTwoNumbers:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        result = ListNode(0)
        temp = result
        carry = 0
        while l1 is not None or l2 is not None:
            temp.next = ListNode(carry)
            temp = temp.next
            if l1 is not None:
                temp.val += l1.val
                l1 = l1.next
            if l2 is not None:
                temp.val += l2.val
                l2 = l2.next
            carry = temp.val // 10
            temp.val %= 10
        if carry != 0:
            temp.next = ListNode(carry)
        return result.next

File: Go/medium/AddTwoNumbers.go
package medium

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	// new函数会分配内存，返回的值是一个指向该类型零值的地址
	head := new(ListNode)
	cur := head
	carry := 0
	for l1 != nil || l2 != nil {
		if l1 != nil {
			carry += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			carry += l2.Val
			l2 = l2.Next
		}
		cur.Next = &ListNode{
			Val:  carry % 10,
			Next: nil,
		}
		carry /= 10
		cur = cur.Next
	}
	if carry != 0 {
		cur.Next = &ListNode{
			Val:  1,
			Next: nil,
		}
	}
	return head.Next
}
